scalableminds/webknossos
# Generate agglomerate skeletons from voxelytics agglomerate file

Generate skeletons from an activated voxelytics agglomerate file mapping by `shift` middle-clicking onto a cell of a segmentation layer.

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### TODO
 - [x] readable error toasts (from backend)
 - [x] antd messages about having sent a request ”loading skeleton for agglomerate 123”
 - [x] frontend eats the error message? (seeing empty toast)
 - [x] error message if clicked with shortcut but action is not available (e.g. no agglomerate mapping was selected)
 - [x] measure time in backend
 - [x] if “render missing data black” is deactivated, clicking on rendered mag2 if zoomed in does not send requests.
 - [x] decide if skeletons should be saved in annotation (or should this entire thing be a view-mode/sandbox feature?)
 - [x] do not load isosurface on ctrl+shift click, even if it is active in the settings
 - [x] should be able to request an agglomerate again if selected mapping changed
 - [x] skeletons can be loaded multiple times, currently - kind of related to "decide if skeletons should be saved in annotation ..." - _We decided that this is not a problem, the user will notice that there are duplicated trees_
 - [x] if requests fail, progress message should disappear, should be able to try again
 - [x] re-enable full tests before merging
 - [x] send request to fallback layer instead of volume annotation layer for volume/hybrid annotations
 - [x] show error if there is a volume annotation layer but no fallback layer -> _solved because there is no agglomerate file mapping in that case_
 - [x] find new shortcut, `ctrl + shift + leftclick` is already taken for the fill tool :see_no_evil: 
 - [x] document keyboard shortcut _Shift + Middle Click_

### Steps to test:
- Locally, use a dataset with agglomerate file mappings (for example, by linking the voxelytics test output). If you don't have one, I can send a test dataset.
- Activate a mapping and `shift + middle click` onto a cell -> Progress notifications should be shown and a skeleton should be loaded.
- Test all the error cases:
  - without activating a mapping
  - with an activated json mapping instead of hdf5
  - clicking on the background (segment 0)
- Check the TODOs above for further test cases

### Issues:
- fixes #4886 

------
